examples/joker-middleware-example/README.md

```diff
@@ -0,0 +1,106 @@
+might crash the stream.
+
+Let's start with the `stream` code:
+
+```python
+@stream_engine.stream(topics=["local--hello-world"])
+async def consume(cr: kstreams.ConsumerRecord):
+    print(f"Event consumed: headers: {cr.headers}, payload: {cr}")
+    if cr.value == b"joker":
+        raise ValueError("Stream crashed 🤡 🤡 🤡 🤡 🤡 🤡 🤡 🤡")
+```
+
+If an event with `joker` then we have the follwing error, causing a hard crash:
+
+```bash
+Event consumed: headers: (), payload: ConsumerRecord(topic='local--hello-world', partition=0, offset=2, timestamp=1704453280808, timestamp_type=0, key=None, value=b'joker', checksum=None, serialized_key_size=-1, serialized_value_size=5, headers=())
+Stream consuming from topics ['local--hello-world'] CRASHED!!! 
+
+ Stream crashed 🤡 🤡 🤡 🤡 🤡 🤡 🤡 🤡
+Traceback (most recent call last):
+  File "Projects/kstreams/kstreams/middleware/middleware.py", line 60, in __call__
+    await self.next_call(cr)
+  File "Projects/kstreams/kstreams/streams.py", line 313, in __call__
+    await self.handler(cr)
+  File "Projects/kstreams/examples/middleware-example/middleware_example/app.py", line 22, in consume
+    raise ValueError("Stream crashed 🤡 🤡 🤡 🤡 🤡 🤡 🤡 🤡")
+ValueError: Stream crashed 🤡 🤡 🤡 🤡 🤡 🤡 🤡 🤡
+```
+
+How can we solve the problem?
+
+- Write buisness logic inside the `stream` to handle the `ValueError` exception
+- Create a new funcion that receives the `cr` as parameter. It will also handle the `ValueError` exception, then all the buiness logic can reused and the stream will be simpler
+- Create a `middleware` with the logic to handle the exception and attach it to the `stream` or the `stream_engine`
+
+For this case we will create a `middleware` just to show you the concept. In real use cases, this middleware could represent a `Dead Letter Queue` (DLQ) for example
+
+```python
+class JokerBlokerMiddleware(kstreams.middleware.BaseMiddleware):
+    async def __call__(self, cr: kstreams.ConsumerRecord) -> None:
+        try:
+            await self.next_call(cr)
+        except ValueError:
+            print("Blocking the Joker 🤡⃤ 🤡⃤ 🤡⃤. so we can continue processing events... \n")
+
+middlewares = [kstreams.middleware.MiddlewareFactory(JokerBlokerMiddleware)]
+
+
+@stream_engine.stream(topics=["local--hello-world"], middlewares=middlewares)
+async def consume(cr: kstreams.ConsumerRecord):
+    print(f"Event consumed: headers: {cr.headers}, payload: {cr}")
+    if cr.value == b"joker":
+        raise ValueError("Stream crashed 🤡 🤡 🤡 🤡 🤡 🤡 🤡 🤡")
+```
+
+Now that we have applied the `JokerBlokerMiddleware`, if we receive the `joker` event the `stream` will not crash!!!
+
+```bash
+Event consumed: headers: (), payload: ConsumerRecord(topic='local--hello-world', partition=0, offset=9, timestamp=1704453577436, timestamp_type=0, key=None, value=b'joker', checksum=None, serialized_key_size=-1, serialized_value_size=5, headers=())
+Blocking the Joker 🤡⃤ 🤡⃤ 🤡⃤. so we can continue processing events... 
```

Comment:
I'm a bit confused how this works..

1. I understand the middleware happens right away.
2. What's weird, is I would think since you caught the exception, it would continue and still print stream crashed.. 

Can you clarify how this works? we can chat about it if necessary

UPDATE: Perhaps I understand now.. The middleware is catching the value error thrown by the consume function.. very interesting..

Reply:
Exactly!